learning/01_Exercises/Lab 7 to 9/UI/userInterface.py

```diff
@@ -0,0 +1,486 @@
+        for theClient, theCount in self.controller.clientsRepository.trackClientRentedMovies.items():
+            print(theClient, theCount)
+
+    def displayTop30Clients(self, clients):
+        for client, count in clients.items():
+            print(client, count)
+
+    def addMovie(self):
+        while True:
+            # Refresh the screen.
+            self.refreshScreen()
+
+            # Display the title and the current list with movies.
+            print("ADD MOVIES")
+            self.displayMovies()
+
+            # Collect a new movie from the user.
+            newMovie = self.collectMovieData()
+
+            # Try/add the movie to the movie list/movieRepo.
+            self.controller.addMovie(newMovie)
+
+            # Ask the user if he wants to continue or not.
+            if not self.exit():
+                break
+
+    def addClients(self):
+        while True:
+            # Refresh the screen.
+            self.refreshScreen()
+
+            # Display the title and the current list with clients.
+            print("ADD CLIENTS")
+            self.displayClients()
+
+            # Collect a new client from the user.
+            Client = self.collectClientsData()
+
+            # Try/add the client to the client list/clientRepo.
+            self.controller.addClient(Client)
+
+            # Ask the user if he wants to continue or not.
+            if not self.exit():
+                break
+
+    def removeMovies(self):
+        while True:
+            # Refresh the screen.
+            self.refreshScreen()
+
+            # Display the title and the current list with clients.
+            print("REMOVE MOVIES")
+            self.displayMovies()
+
+            # Collect the movie ID.
+            ID = self.collectID()
+
+            # Try/remove the movie with the ID given.
+            self.controller.removeMovie(ID)
+
+            # Ask the user if he wants to continue
+            if not self.exit():
+                break
+
+    def removeClient(self):
+        while True:
+            # Refresh the screen.
+            self.refreshScreen()
+
+            # Display the title and the current list with clients.
+            print("REMOVE CLIENTS")
+            self.displayClients()
+
+            # Collect the movie ID.
+            ID = self.collectID()
+
+            # Try/remove the movie with the ID given.
+            self.controller.removeClient(ID)
+
+            # Ask the user if he wants to continue
+            if not self.exit():
+                break
+
+    def updateMovies(self):
+        while True:
+            # Refresh the screen.
+            self.refreshScreen()
+
+            # Display the title and the current list with clients.
+            print("UPDATE MOVIES")
+            self.displayMovies()
+
+            # Collect the movie ID.
+            ID = self.collectID()
+            if not self.controller.validatorForMovie.checkID(ID, self.controller.moviesRepository.getIdList()):
```

Comment:
There is a problem with the way you are calling these validators, more specifically the fact that you are passing every time a list of ids. These makes me raise my eyebrows since a validator responsability is to validate that an object is valid from a logical stand point. In my opinion checking if the object is already in the database should be done in controller ( or repository should raise an erro if the object with the corresponding id already exists)

Reply:
Rezolvat.